lib/dialects/abstract/index.ts

```diff
@@ -0,0 +1,142 @@
+type DialectSupports = {
+  'DEFAULT': boolean,
+  'DEFAULT VALUES': boolean,
+  'VALUES ()': boolean,
+  'LIMIT ON UPDATE': boolean,
+  'ON DUPLICATE KEY': boolean,
+  'ORDER NULLS': boolean,
+  'UNION': boolean,
+  'UNION ALL': boolean,
+  'RIGHT JOIN': boolean,
+
+  /* does the dialect support returning values for inserted/updated fields */
+  returnValues: boolean,
+
+  /* features specific to autoIncrement values */
+  autoIncrement: {
+    /* does the dialect require modification of insert queries when inserting auto increment fields */
+    identityInsert: boolean,
+
+    /* does the dialect support inserting default/null values for autoincrement fields */
+    defaultValue: boolean,
+
+    /* does the dialect support updating autoincrement fields */
+    update: boolean
+  },
+  /* Do we need to say DEFAULT for bulk insert */
+  bulkDefault: boolean,
+  schemas: boolean,
+  transactions: boolean,
+  settingIsolationLevelDuringTransaction: boolean,
+  transactionOptions: {
+    type: boolean
+  },
+  migrations: boolean,
+  upserts: boolean,
+  inserts: {
+    ignoreDuplicates: string, /* dialect specific words for INSERT IGNORE or DO NOTHING */
+    updateOnDuplicate: boolean, /* whether dialect supports ON DUPLICATE KEY UPDATE */
+    onConflictDoNothing: string, /* dialect specific words for ON CONFLICT DO NOTHING */
+    conflictFields: boolean /* whether the dialect supports specifying conflict fields or not */
+  },
+  constraints: {
+    restrict: boolean,
+    addConstraint: boolean,
+    dropConstraint: boolean,
+    unique: boolean,
+    default: boolean,
+    check: boolean,
+    foreignKey: boolean,
+    primaryKey: boolean
+  },
+  index: {
+    collate: boolean,
+    length: boolean,
+    parser: boolean,
+    concurrently: boolean,
+    type: boolean,
+    using: boolean,
+    functionBased: boolean,
+    operator: boolean
+  },
+  groupedLimit: boolean,
+  indexViaAlter: boolean,
+  JSON: boolean,
+  deferrableConstraints: boolean
+};
+
+export class AbstractDialect {
+  // the actual field is on the prototype,
+  //  hence the `declare` keyword
+  declare supports: DialectSupports;
+
+  // TODO: Replace with QueryGenerator class once its typings are complete.
+  declare queryGenerator: unknown;
+}
+
+AbstractDialect.prototype.supports = {
```

Comment:
Any reason why we don't inline this?

```
export class AbstractDialect {
  const supports: DialectSupports = { /* stuff here */ } 
}
```

Reply:
I copied this part over but my guess is the initial idea was to be able to access this object without instantiating the Dialect class.

Reply:
What makes more sense from your pov?

Reply:
I think adding a field to a prototype is prone to errors. I'd prefer assigning a frozen object to the constructor & the instance.

Let's find out if anything breaks if we do this.

Reply:
Done. `AbstractDialect.prototype.supports` still exists but it's now a getter that returns the static field

Reply:
I was about to say, we should refactor the usages of protoype out as much as possible in later PRs.